Documented connection hints as well as keep-alives within all protocol states (#16)
neo4j-drivers/neo4j-drivers.github.io: bolt/bolt-protocol-message-specification-4.md

No changes compared to version 4.1.

# Version 4.3

## Deltas

The changes compared to Bolt protocol version 4.2 are listed below:

 * `NOOP` chunks may now be transmitted in all connection states when a connection remains in idle for extended
   periods of time while the server is busy processing a request.
 * An additional `hint` dictionary has been added to the `metadata` property of the `SUCCESS` structure transmitted in
   response to the `HELLO` command in order to provide optional configuration hints to drivers.

### Request Message - 4.3 - `HELLO`

> The `HELLO` message request the connection to be authorized for use with the remote database.

Reviewer: request*s*?

**Detail Messages:**

No detail messages should be returned.

**Valid Summary Messages:**

* `SUCCESS`
* `FAILURE`

#### Synopsis

```
HELLO {user_agent::String, scheme::String, principal::String, credentials::String, routing::Dictionary(address::String))
```

Example 1:

```
HELLO {"user_agent": "Example/4.1.0", "scheme": "basic", "principal": "user", "credentials": "password", "routing": {"address": "x.example.com:9001"}}
```

Example 2:

```
HELLO {"user_agent": "Example/4.1.0", "scheme": "basic", "principal": "user", "credentials": "password", "routing": {"address": "x.example.com:9001", "policy": "example_policy_routing_context", "region": "example_region_routing_context"}}
```

#### Server Response `SUCCESS`

A `SUCCESS` message response indicates that the client is permitted to exchange further messages.
Servers can include metadata that describes details of the server environment and/or the connection.

The following fields are defined for inclusion in the `SUCCESS` metadata.

- `server::String` (server agent string, example `"Neo4j/4.1.0"`)
- `connection_id::String` (unique identifier of the bolt connection used on the server side, example: `"bolt-61"`)

> - `hints::Dictionary` (set of optional configuration hints to be considered by the driver )

Reviewer: space before `)`

The new `hints` dictionary may contain a set of optional configuration hints which may be interpreted or ignored by
drivers at their own discretion in order to augment operations where applicable. A full listing of the available
hints may be found in [Appendix B](#appendix-b---connection-hints).

Example:

```

> SUCCESS {"server": "Neo4j/4.0.0", "hints": {"connection.recv_timeout_seconds":120}}

Reviewer: `{"connection.recv_timeout_seconds": 120}`
(added space)

No changes compared to version 4.1.

# Version 4.3

## Deltas

The changes compared to Bolt protocol version 4.2 are listed below:

 * `NOOP` chunks may now be transmitted in all connection states when a connection remains in idle for extended
   periods of time while the server is busy processing a request.

>  * An additional `hint` dictionary has been added to the `metadata` property of the `SUCCESS` structure transmitted in

Reviewer 2: ```suggestion
 * An additional `hints` dictionary has been added to the `metadata` property of the `SUCCESS` structure transmitted in
```

C: COMMIT
S: SUCCESS {"bookmark": "neo4j:bookmark-test-1"}
```

### Appendix B - Connection Hints

As of version 4.3, the server may optionally include a set of configuration hints within the `SUCCESS` metadata provided
in response to the `HELLO` message. Drivers may choose to interpret or ignore these hints at their own discretion. The
following set of hints are currently defined:

#### connection.recv_timeout_seconds - 4.3

> Maximum amount of time (in seconds) for which a connection may remain in idle following a request before drivers should

Reviewer 2: ```suggestion
Maximum amount of time (in seconds) for which a connection may remain idle following a request before drivers should
```

Reviewer 2: might be worth simplifying the sentence a bit

consider it stale.

> The server will ensure that a chunk (or `NOOP` chunk) will be transmitted at least once per timeout period. Drivers

Reviewer 2: ```suggestion
The server will ensure that a chunk (like a `NOOP` chunk e.g.) will be transmitted at least once per timeout period. Drivers
```

> which elect to interpret this hint may terminate connections which are considered stale by the definition of this hint.

Reviewer 2: ```suggestion
which choose to interpret this hint may terminate connections which are considered stale by the definition of this hint.
```
